sky/jobs/log_gc.py

```diff
@@ -0,0 +1,187 @@
+            break
+        except Exception as e:  # pylint: disable=broad-except
+            logger.error(f'Error GC controller logs for job: {e}',
+                         exc_info=True)
+        interval = _next_gc_interval(controller_logs_retention)
+        logger.info('Next controller logs GC is scheduled after '
+                    f'{interval} seconds')
+        await asyncio.sleep(interval)
+
+
+async def gc_task_logs_for_job():
+    """Garbage collect task logs for job."""
+    while True:
+        skypilot_config.reload_config()
+        task_logs_retention = skypilot_config.get_nested(
+            ('jobs', 'controller', 'task_logs_gc_retention_hours'),
+            _DEFAULT_TASK_LOGS_GC_RETENTION_HOURS) * 3600
+        logger.info(
+            f'GC task logs for job: retention {task_logs_retention} seconds')
+        if task_logs_retention < 0:
+            continue
+        try:
+            finished = False
+            while not finished:
+                finished = await _clean_task_logs_with_retention(
+                    task_logs_retention)
+        except asyncio.CancelledError:
+            logger.info('Task logs GC task cancelled')
+            break
+        except Exception as e:  # pylint: disable=broad-except
+            logger.error(f'Error GC task logs for job: {e}', exc_info=True)
+        interval = _next_gc_interval(task_logs_retention)
+        logger.info(f'Next task logs GC is scheduled after {interval} seconds')
+        await asyncio.sleep(_next_gc_interval(task_logs_retention))
+
+
+async def _clean_controller_logs_with_retention(retention_seconds: int,
+                                                batch_size: int = 1000):
+    """Clean controller logs with retention.
+
+    Returns:
+        Whether the GC of this round has finished, False means there might
+        still be more controller logs to clean.
+    """
+    assert batch_size > 0, 'Batch size must be positive'
+    jobs = await managed_job_state.get_controller_logs_to_clean_async(
+        retention_seconds, batch_size=batch_size)
+    for job in jobs:
+        log_file = managed_job_utils.controller_log_file_for_job(job['job_id'])
+        if os.path.exists(log_file):
+            cleaned_at = time.time()
+            ts_str = datetime.fromtimestamp(cleaned_at).strftime(
+                '%Y-%m-%d %H:%M:%S')
+            msg = f'Controller log has been cleaned at {ts_str}.'
+            # Sync down logs will reference to this file directly, so we
+            # keep the file and delete the content.
+            # TODO(aylei): refactor sync down logs if the inode usage
+            # becomes an issue.
+            async with await anyio.open_file(log_file, 'w',
+                                             encoding='utf-8') as f:
+                await f.write(msg + '\n')
+        await managed_job_state.set_controller_logs_cleaned_async(
+            job_id=job['job_id'], logs_cleaned_at=cleaned_at)
```

Comment:
Can we batch these queries? I'm worried about the case where we end up making thousands of SQLite DB writes in quick succession, which may cause some locking issues if there are a lot of running jobs at the same time.